tarball/tarball_transport.go

```diff
@@ -0,0 +1,64 @@
+package tarball
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strings"
+
+	"github.com/containers/image/transports"
+	"github.com/containers/image/types"
+)
+
+const (
+	transportName = "tarball"
+	separator     = ":"
+)
+
+var (
+	// Transport implements the types.ImageTransport interface for "tarball:" images,
+	// which are makeshift images constructed using one or more possibly-compressed tar
+	// archives.
+	Transport = &tarballTransport{}
+)
+
+type tarballTransport struct {
+}
+
+func (t *tarballTransport) Name() string {
+	return transportName
+}
+
+func (t *tarballTransport) ParseReference(reference string) (types.ImageReference, error) {
+	var stdin []byte
+	var err error
+	filenames := strings.Split(reference, separator)
+	for _, filename := range filenames {
+		if filename == "-" {
+			stdin, err = ioutil.ReadAll(os.Stdin)
```

Comment:
We need the tarball contents in order to compute the blobsum to include in the manifest (after #305, we can shorten that up a bit by using the `manifest` package rather than building the structure from scratch), and again in GetBlob(), conceivably via multiple ImageSource objects built by the ImageReference.  We can't reasonably expect to be able to rewind stdin to read the contents again, so we have to buffer it somewhere.

We buffer it in the reference object so that we can modify the manifest there, before handing it off to copy.Image(), and still get a usable result.

Buffering to a file in the reference object is messy because it doesn't have an explicit Close() method, and between that and needing to be able to create zero or one or more ImageSource objects from the reference, managing the file's lifetime got... messy.  I'd like to use a file, but could use some advice on how best to manage that.

Reply:
((

> We need the tarball contents in order to compute the blobsum to include in the manifest …

Sure, but that computation first happens in `NewImageSource`, not before. Couldn’t the stdin be opened and buffered at that point as well, instead of in here?

`ImageSource` has a clearly delineated lifetime (between `NewImageSource` and `Close`), which allows creating and deleting temporary files; `ImageReference` is expected to be ~immutable and doesn’t have a `Close` operation, forcing the code to use memory.

> We buffer it in the reference object so that we can modify the manifest there

I can’t see how the two interact; `ParseReference` stores the list of file names into a private member, so the caller must call `ConfigUpdate` without taking advantage of any information from `ref.stdin`. If `ref.stdin` were not computed at that time, the caller of `ConfigUpdate` can’t tell the difference AFAICT.

))

---

> between that and needing to be able to create zero or one or more ImageSource objects from the reference,

OK, _this_ is a strong argument for not reading `stdin` in `NewImageSource` — but what in particular needs to create more than one `ImageSource` from a single reference? `copy.Image` does not do that at least.

(I guess the caller of all of this could always replace `"-"` by its own temporary file; e.g. `kpod load` does that IIRC. But again, if we are expecting the caller to do something specific to the `tarball:` transport, we could just as well have it call `CreateImageFromConfigAndTarballs`.)

Reply:
A `CreateImageFromConfigAndTarballs` wouldn't produce something that could be used as a source for copy.Image(), though, and having to ask it for a reference that would then be used to create another ImageSource makes the file management complicated again.

Reply:
> A `CreateImageFromConfigAndTarballs` wouldn't produce something that could be used as a source for copy.Image(), though

Right, that would not be usable for `skopeo copy tarball:$path $dest`, but a `kpod import $tarball_path $dest` would be possible.

Reply:
FWIW, I was hoping to be able to use this as a source for `buildah from` as well, though admittedly it's not much faster than running `buildah from scratch` and then using `buildah add` to add tarball contents.

I would have expected its use of the `tarball` transport to remain an internal detail of `kpod import`.

Reply:
I dunno; this _does_ work, and the `stdin` ugliness is very encapsulated and does not leak anywhere else.  So I can’t really object to this PR.

It just seems to me that the ugliness is mostly self-imposed, somehow (by an inappropriate `ImageSource` API? by `copy.Image` rigid about sources/destinations and somehow not reusable?) and that there should be a cleaner way to organize $something. I don’t really have an idea how to fix things, though.